### PR TITLE
Chapter 5 exercise 3 Everything KAN-37

### PR DESCRIPTION
**Pull request by insert-oreyes**
- Add component WeatherApp with the creative result, documentation and testing
- Update chapter 5 page, render exercise 3 and testing
- Update db with exercise 3 details

### CODE REVIEW

File: components/exercises/chapter-5-oscar/Everything/every.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/every.tsx]: why limit only to numbers? I think this function should be able to handle arrays of any type

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.test.tsx]: Can we add tests for arrays that contain other elements apart from numbers? Like strings, objects, arrays, null, undefined....

File: components/exercises/chapter-5-oscar/Everything/every.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/every.test.tsx]: Can we add tests for arrays that contain other elements apart from numbers? Like strings, objects, arrays, null, undefined....

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.test.tsx]: it('should return true if at least one element satisfies the condition', () => {

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.test.tsx]: it('should return false if none of the element satisfies the condition', () => {

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.test.tsx]: there seems to be a bug here, since 2, 4,7 and 8 are < 10, this call should return true

File: components/exercises/chapter-5-oscar/Everything/some.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.tsx]: I don't think you are allowed to use `Array.some` here, the point of this exercise is to implement your own version using a loop

Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/some.tsx]: literally says "using the some method"

File: components/exercises/chapter-5-oscar/Everything/styles.css
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/styles.css]: The problem with adding a style like this is that, if in any other component we add a className of `wrapper`, it will also be affected by this styling. So it doesn't affect only to the component, but is added to the global styling namespace. This can cause styling bugs that are really difficult to track.

To make sure this styling is encapsulated to this component, can we replace this by some tailwind classes? or if it can't be done with tailwind, can we use inline styles like `<div style={{ opacity: 1 }}>`?

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/some.test.tsx]: It hits false because 11 is higher than 10. I am evaluating the Boolean result as false as if it returns true, the function is wrong

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx:21]: Can you add a comment like "This key is free and public so there's no harm on leaving it right here"

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: Why axios? can't we just use `fetch`?

File: package.json
Comment (jportella93) [inline package.json]: Why next upgrade?

File: package.json
Comment (jportella93) [inline package.json]: Why are we adding this? In general, installing a new dependency requires a justification. Think the more dependencies a project has, the higher number of things are out of our control, so we need to add them sparingly.

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: Can we add a type for the error? In general, only use `any` if there's absolutely no other way

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: if (error.response?.status === 404) {

File: package.json
Comment (insert-oreyes) [reply package.json]: I needed to add this dependency to use axios

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: {weatherData.name && (

File: package.json
Comment (insert-oreyes) [reply package.json]: Because I had issues with tailwind classes and phind suggested to update next

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: Because I never have used it before and I thought it would be a good opportunity to try

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: What does this do? This logic is a bit hard to follow, some comments or making it a function above with a descriptive name and calling it here would help

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:84]: can't we bundle this 3 `await waitFor(() => {`?

Like
```js
await waitFor(() => {
      const cityName = screen.getByText('SampleCity');
      expect(cityName).toBeInTheDocument();
      const temperature = screen.getByText('22 ºC');
      expect(temperature).toBeInTheDocument();
```

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: I see we are repeating these 4 lines over and over. Might be worth adding an util above, so we can be more concise on tests

```js
submitFormWithCity('SampleCity')
```

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:117]: Can we test the error isn't shown before submitting the form?

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:141]: Can we test the error isn't shown before submitting the form?

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: This mock data seems also copy pasted from the first test, can we declare it where it is in scope of all tests? So we don't need to copy paste it in each test

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:84]: if I do, the code complains: Avoid using multiple assertions within `waitFor` callback

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: This can be just added to the first test instead of creating a new whole test for it

File: components/exercises/chapter-5-oscar/Everything/every.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/every.tsx]: I limited it to numbers taking as reference the exercise outcome and also numbers is what I am using it for

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/some.test.tsx]: should return false if any of the element does not satisfy the condition

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/some.test.tsx]: refactored

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx:21]: added, thank you

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: refactored, thank you

File: package.json
Comment (insert-oreyes) [reply package.json]: I have downgrade it as it was

File: package.json
Comment (insert-oreyes) [reply package.json]: removed

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: added in the state

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: refactored

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: refactored

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: refactored, thank you

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:117]: Added

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx:141]: added

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: extracted out of scope, thank you

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.test.tsx]: added to the first test

File: components/exercises/chapter-5-oscar/Everything/every.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/every.tsx]: I've got the following error if I add strings aside numbers: 
The error "Operator '>' cannot be applied to types 'string | number' and 'number'.ts(2365)". Therefore the solution was TypeScript's generic type parameter (<T>) to make the function more flexible and allow it to work with arrays containing elements of different types.

File: components/exercises/chapter-5-oscar/Everything/some.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/some.test.tsx]: test added for every case

File: components/exercises/chapter-5-oscar/Everything/every.test.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/every.test.tsx]: test added for every case

File: components/exercises/chapter-5-oscar/Everything/styles.css
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/styles.css]: I fixed like below, I had to randomly generate images as I couldn't a local file:
style={{backgroundImage: `url(https://source.unsplash.com/1600x900/?${backgrounfPhoto})`, opacity: 0.5,}}

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (insert-oreyes) [reply components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: Sorry, to make it clear I refactored it. It compares if the feel_like temperature is higher than the temp_max forecasted. Therefore, if it is higher, it will display "🔥 The temperature feels warmer than the maximum" else "👍 The temperature does not feel over the maximum"

File: components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/WeatherApp.tsx]: axios?

File: components/exercises/chapter-5-oscar/Everything/some.test.ts
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.test.ts]: Nope, `some` should return true **if at least one element satisfies the condition**

Let's update the function code so this test passes
```js
    const arr = [20, 4, 7, 8];
    const callBack = (num: number) => num < 10;
    expect(some(arr, callBack)).toBe(true);
```

Comment (jportella93) [reply components/exercises/chapter-5-oscar/Everything/some.test.ts]: Actually forget about this, read my comment below

File: components/exercises/chapter-5-oscar/Everything/some.ts
Comment (jportella93) [inline components/exercises/chapter-5-oscar/Everything/some.ts:1]: I think there's a confusion with the exercise requirements

>Implement every as a function that takes an array and a predicate function as parameters. **Write two versions**, one using a loop and one using the some method.

It is asking to write to versions of `every`, one of them using `some`. Not write a function called `some`.

So you already have one `every` working, for the other one you could call it `everyWithSome` and the tests should be the same as for `every`

```js
expect(everyWithSome([1, 2, 3], (n) => n < 10)).to.beTrue()
```

**Review by jportella93 — APPROVED** [review]
Approving to unblock but let's address the comments before merging

**Pull request merged**
